src/mempoollimit.cpp

```diff
@@ -0,0 +1,159 @@
+
+#include "core_memusage.h"
+#include "mempoollimit.h"
+#include "random.h"
+#include "serialize.h"
+#include "timedata.h"
+#include "version.h"
+
+const CAmount DEFAULT_FEE = 10000;
+const TxWeight ZERO_WEIGHT = TxWeight(0, 0);
+
+void RecentlyEvictedList::pruneList()
+{
+    if (txIdSet.empty()) {
+        return;
+    }
+    int64_t now = GetAdjustedTime();
+    size_t startIndex = (txIdsAndTimesIndex + capacity - txIdSet.size()) % capacity;
+    boost::optional<std::pair<uint256, int64_t>> txIdAndTime;
+    while ((txIdAndTime = txIdsAndTimes[startIndex]).is_initialized() && (now - txIdAndTime.get().second) > timeToKeep) {
+        txIdsAndTimes[startIndex] = boost::none;
+        txIdSet.erase(txIdAndTime.get().first);
+        startIndex = (startIndex + 1) % capacity;
+    }
+}
+
+void RecentlyEvictedList::add(const uint256& txId)
+{
+    pruneList();
+    if (txIdsAndTimes[txIdsAndTimesIndex].is_initialized()) {
+        auto txIdAndTime = txIdsAndTimes[txIdsAndTimesIndex];
+        txIdSet.erase(txIdAndTime.get().first);
+    }
+    txIdsAndTimes[txIdsAndTimesIndex] = std::make_pair(txId, GetAdjustedTime());
+    txIdSet.insert(txId);
+    txIdsAndTimesIndex = (txIdsAndTimesIndex + 1) % capacity;
+}
+
+bool RecentlyEvictedList::contains(const uint256& txId)
+{
+    pruneList();
+    return txIdSet.count(txId) > 0;
+}
+
+
+TxWeight WeightedTxTree::getWeightAt(size_t index) const
+{
+    return index < size ? txIdAndWeights[index].txWeight.add(childWeights[index]) : ZERO_WEIGHT;
+}
+
+void WeightedTxTree::backPropagate(size_t fromIndex, const TxWeight& weightDelta)
+{
+    while (fromIndex > 0) {
+        fromIndex = (fromIndex - 1) / 2;
+        childWeights[fromIndex] = childWeights[fromIndex].add(weightDelta);
+    }
+}
+
+size_t WeightedTxTree::findByWeight(size_t fromIndex, uint64_t weightToFind) const
+{
+    int leftWeight = getWeightAt(fromIndex * 2 + 1).lowFeePenaltyWeight;
+    int rightWeight = getWeightAt(fromIndex).lowFeePenaltyWeight - getWeightAt(fromIndex * 2 + 2).lowFeePenaltyWeight;
+    // On Left
+    if (weightToFind < leftWeight) {
+        return findByWeight(fromIndex * 2 + 1, weightToFind);
+    }
+    // Found
+    if (weightToFind < rightWeight) {
+        return fromIndex;
+    }
+    // On Right
+    return findByWeight(fromIndex * 2 + 2, weightToFind - rightWeight);
+}
+
+TxWeight WeightedTxTree::getTotalWeight() const
+{
+    return getWeightAt(0);
+}
+
+
+void WeightedTxTree::add(const WeightedTxInfo& weightedTxInfo)
+{
+    txIdAndWeights.push_back(weightedTxInfo);
+    childWeights.push_back(ZERO_WEIGHT);
+    txIdToIndexMap[weightedTxInfo.txId] = size;
+    backPropagate(size++, weightedTxInfo.txWeight);
+}
+
+void WeightedTxTree::remove(const uint256& txId)
+{
+    if (txIdToIndexMap.find(txId) == txIdToIndexMap.end()) {
+        return;
+    }
+
+    size_t removeIndex = txIdToIndexMap[txId];
+
+    TxWeight lastChildWeight = txIdAndWeights[--size].txWeight;
```

Comment:
`size` should be decremented explicitly, not as a side-effect of this line.

Reply:
I personally like inlining `++` and `--` where I can, but I will change this for readability.